From 6723f1a54e0e7c13b1506f2f30b73c84ac4920fc Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Tue, 31 Jul 2018 13:45:48 +0800
Subject: [PATCH 16/51] soc: ti: add HWMOD_INIT_NO_IDLE to workaround the
 hardware issue

On am3358, there is a bug in silicon that clocks cannot be cut to DEBUGSS.
It has been confirmed from the following vendor website:

https://e2e.ti.com/support/arm/sitara_arm/f/791/p/713673/2632160#2632160

A kernel fragment AM335X_DEBUGSS_NO_IDLE is introduced to workaround it
in case this hardware issue might be OK in the latest version. Currently
it is enabled by default in scc/cfg.

Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
(cherry picked from commit a8cf9b9cdd1267fefe9332224278f622032f14eb)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 arch/arm/mach-omap2/omap_hwmod_33xx_data.c | 3 +++
 drivers/soc/ti/Kconfig                     | 7 +++++++
 2 files changed, 10 insertions(+)

diff --git a/arch/arm/mach-omap2/omap_hwmod_33xx_data.c b/arch/arm/mach-omap2/omap_hwmod_33xx_data.c
index 41d101494d8d..932dc8bfdac5 100644
--- a/arch/arm/mach-omap2/omap_hwmod_33xx_data.c
+++ b/arch/arm/mach-omap2/omap_hwmod_33xx_data.c
@@ -202,6 +202,9 @@ static struct omap_hwmod am33xx_debugss_hwmod = {
 	.name		= "debugss",
 	.class		= &am33xx_debugss_hwmod_class,
 	.clkdm_name	= "l3_aon_clkdm",
+#ifdef CONFIG_AM335X_DEBUGSS_NO_IDLE
+	.flags		= HWMOD_INIT_NO_IDLE,
+#endif
 	.main_clk	= "trace_clk_div_ck",
 	.prcm		= {
 		.omap4	= {
diff --git a/drivers/soc/ti/Kconfig b/drivers/soc/ti/Kconfig
index d7d50d48d05d..88251dc1e03f 100644
--- a/drivers/soc/ti/Kconfig
+++ b/drivers/soc/ti/Kconfig
@@ -75,6 +75,13 @@ config TI_SCI_PM_DOMAINS
 	  called ti_sci_pm_domains. Note this is needed early in boot before
 	  rootfs may be available.
 
+config AM335X_DEBUGSS_NO_IDLE
+	bool "AM335X DEBUGSS no idle"
+	depends on SOC_AM33XX
+	help
+	  Enable this item if your soc has a bug in silicon that clocks can
+	  not be cut to DEBUGSS. Otherwise, disable it.
+
 endif # SOC_TI
 
 config TI_SCI_INTA_MSI_DOMAIN
-- 
2.17.1

